code/game/objects/structures/tank_dispenser.dm

```diff
@@ -88,40 +94,55 @@
 
 /obj/structure/dispenser/Topic(href, href_list)
 	if(..())
-		return 1
+		return TRUE
 
 	if(Adjacent(usr))
 		usr.set_machine(src)
+
+		// The oxygen tank button
 		if(href_list["oxygen"])
-			if(oxygentanks > 0)
-				var/obj/item/tank/oxygen/O
-				if(oxytanks.len == oxygentanks)
-					O = oxytanks[1]
-					oxytanks.Remove(O)
-				else
-					O = new /obj/item/tank/oxygen(loc)
-				O.loc = loc
-				to_chat(usr, "<span class='notice'>You take [O] out of [src].</span>")
-				oxygentanks--
-				update_icon()
+			try_remove_tank(usr, stored_oxygen_tanks)
+
+		// The plasma tank button
 		if(href_list["plasma"])
-			if(plasmatanks > 0)
-				var/obj/item/tank/plasma/P
-				if(platanks.len == plasmatanks)
-					P = platanks[1]
-					platanks.Remove(P)
-				else
-					P = new /obj/item/tank/plasma(loc)
-				P.loc = loc
-				to_chat(usr, "<span class='notice'>You take [P] out of [src].</span>")
-				plasmatanks--
-				update_icon()
+			try_remove_tank(usr, stored_plasma_tanks)
+
 		add_fingerprint(usr)
 		updateUsrDialog()
-		SSnanoui.update_uis(src)
+		SSnanoui.try_update_ui(usr, src)
 	else
 		SSnanoui.close_user_uis(usr,src)
-	return 1
+	return TRUE
+
+/// Called when the user clicks on the oxygen or plasma tank UI buttons, and tries to withdraw a tank.
+/obj/structure/dispenser/proc/try_remove_tank(mob/living/user, list/tank_list)
+	if(!tank_list.len)
+		return // There are no tanks left to withdraw.
+
+	var/obj/item/tank/T = tank_list[1]
+	tank_list.Remove(T)
+
+	if(!user.put_in_hands(T))
+		T.forceMove(loc) // If the user's hands are full, place it on the tile of the dispenser.
+
+	to_chat(user, "<span class='notice'>You take [T] out of [src].</span>")
+	update_icon()
+
+/// Called when the user clicks on the dispenser with a tank. Tries to insert the tank into the dispenser, and updates the UI if successful.
+/obj/structure/dispenser/proc/try_insert_tank(mob/living/user, list/tank_list, obj/item/tank/T)
+	if(!tank_list.len >= 10)
```

Comment:
Overlooked this last review. But this is currently a magic number. Best to make a define at the top and make the starting values also use that define. Be sure to also undef it after this file is done.